incubator/hnc/pkg/controllers/object_controller_new.go

```diff
@@ -38,25 +59,306 @@ type ObjectReconcilerNew struct {
+	ul.SetGroupVersionKind(r.GVK)
+	if err := r.List(ctx, ul, client.InNamespace(ns)); err != nil {
+		log.Error(err, "Couldn't list objects")
+		return err
+	}
+
+	// Enqueue all the current objects in the namespace because some of them may have been deleted.
+	for _, inst := range ul.Items {
+		if r.isExcluded(log, &inst) {
+			continue
+		}
+		co := object.Canonical(&inst)
+		co.SetNamespace(inst.GetNamespace())
+		log.Info("Enqueuing existing object for reconciliation", "affected", co.GetName())
+		r.Affected <- event.GenericEvent{Meta: co}
+	}
+
+	// Enqueue local copies of the originals in the ancestors to catch any new or changed objects.
+	r.enqueuePropagatedObjects(ctx, log, ns)
+
 	return nil
 }
 
-func (r *ObjectReconcilerNew) SetupWithManager(mgr ctrl.Manager) error {
+func (r *ObjectReconcilerNew) enqueueDescendants(ctx context.Context, log logr.Logger, src *unstructured.Unstructured) {
+	sns := src.GetNamespace()
+	dns := r.Forest.Get(sns).DescendantNames()
+	for _, ns := range dns {
+		dc := object.Canonical(src)
+		dc.SetNamespace(ns)
+		log.Info("Enqueuing descendant copy", "affected", ns+"/"+src.GetName(), "reason", "The source changed")
+		r.Affected <- event.GenericEvent{Meta: dc}
+	}
+}
+
+func (r *ObjectReconcilerNew) enqueuePropagatedObjects(ctx context.Context, log logr.Logger, ns string) {
+	r.Forest.Lock()
+	defer r.Forest.Unlock()
+
+	// Enqueue local copies of the original objects in the ancestors from forest.
+	o := r.Forest.Get(ns).GetPropagatedObjects(r.GVK)
+	for _, obj := range o {
+		lc := object.Canonical(obj)
+		lc.SetNamespace(ns)
+		log.Info("Enqueuing local copy of the ancestor original for reconciliation", "affected", lc.GetName())
+		r.Affected <- event.GenericEvent{Meta: lc}
+	}
+}
+
+// operate operates the action generated from syncing the object with the forest.
+func (r *ObjectReconcilerNew) operate(ctx context.Context, log logr.Logger, act action, inst, so *unstructured.Unstructured) error {
+	switch act {
+	case unchanging:
+		return nil
+	case deleting:
+		return r.delete(ctx, log, inst)
+	case overwriting:
+		return r.overwrite(ctx, log, inst, so)
+	default:
+		// Generate log for any unset action.
+		log.Info("ACTION UNSET!!")
+		return nil
+	}
+}
+
+func (r *ObjectReconcilerNew) delete(ctx context.Context, log logr.Logger, inst *unstructured.Unstructured) error {
+	log.Info("Deleting obsolete copy")
+	err := r.Delete(ctx, inst)
+	if errors.IsNotFound(err) {
+		log.Info("The obsolete copy doesn't exist, no more action needed")
+		return nil
+	}
+	return err
+}
+
+func (r *ObjectReconcilerNew) overwrite(ctx context.Context, log logr.Logger, inst, so *unstructured.Unstructured) error {
+	ns := inst.GetNamespace()
+	inst = object.Canonical(so)
+	inst.SetNamespace(ns)
+	metadata.SetLabel(inst, api.LabelInheritedFrom, so.GetNamespace())
+	log.Info("Overwriting", "dst", inst.GetNamespace(), "origin", so.GetNamespace())
+	err := r.Update(ctx, inst)
+	if err != nil && errors.IsNotFound(err) {
```

Comment:
You can also look and see if `CreationTimestamp()` is zero, and if so call `Create`, else call `Update`. That saves one API call in the (common?) case where we really do need to create.